jax/dtypes.py

```diff
@@ -326,6 +331,11 @@ def dtype(x):
   return np.result_type(x)
 
 def _result_type_raw(*args):
+  """
+  Returns either
+  - a numpy dtype, indicating weak_type=False
+  - a Python type (int, float, complex), indicating weak_type=True
```

Comment:
For both this function and the one above, maybe it'd be better to encode this directly, by making this function return a `(dtype, bool)` pair? That seems like it would reduce risk of surprising the caller, as the caller would have to explicitly handle the weak type tag. WDYT?

Reply:
That's probably a good idea, given how this surprised us!

That said, these return values are indices to the type promotion lattice, so the weak versions of various ints will all have to be normalized to a single identifier somewhere before they're used in the lattice code, and it's not clear where that would be done if not here.

Reply:
How hard would it be to make the lattice operate directly on `(dtype, bool)` pairs?

Reply:
We couldn't use multiple aliases of each weak types in the lattice, because they are unorderable.

Reply:
We could normalize all `(np.u?int*, True)` entries to a single weak int indicator in the lattice, but that's what this function was meant to do from the beginning!